src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/MvcForm.cs

```diff
@@ -52,5 +53,27 @@ protected virtual void Dispose(bool disposing)
                 GenerateEndForm();
             }
         }
+
+        private void RenderEndOfFormContent()
+        {
+            var writer = _viewContext.Writer;
+            var htmlWriter = writer as HtmlTextWriter;
+            var formContext = _viewContext.FormContext;
+            if (formContext.HasRenderAtEndOfFormContent)
+            {
+                foreach (var hiddenTag in formContext.RenderAtEndOfFormContent)
+                {
+                    if (htmlWriter == null)
+                    {
+                        var htmlEncoder = _viewContext.HttpContext.RequestServices.GetRequiredService<IHtmlEncoder>();
```

Comment:
Compiler won't optimize this and will execute the expression every time through the loop. Move this declaration right after `var htmlWriter` and leave it `null` if `htmlWriter` is not `null`.